org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -854,11 +854,30 @@ public static String toUri(IClassFile classFile) {
 			return null;
 		}
 
+		String classId = classFile.getHandleIdentifier();
+		String classFileName = classFile.getElementName();
+		// try to point to the parent class if this is an inner class, this avoid
+		// opening new tab for the actual same source file.
+		if (classFileName.contains("$")) {
+			String parentClassId = classId.substring(0, classId.lastIndexOf("$")) + ".class";
+			IJavaElement parentClass = JavaCore.create(parentClassId);
+			if (parentClass != null && parentClass instanceof IClassFile parentClassFile) {
+				try {
+					// they are from the same source file if the contents are the same.
+					if (Objects.equals(parentClassFile.getBuffer(), classFile.getBuffer())) {

```

Comment:
Regarding your comments on `getBuffer()`, is there ever a case where a nested class would not belong to the same classfile as the part of the element name to the left of the `$` ?

Reply:
Because $ is a valid character for java class name. Given a class named as `Foo$Bar`, we cannot simply tell whether it's class `Bar` nested in `Foo`, or just a class named as `Foo$Bar`.

Though this should be a rare case.

Reply:
I guess the `getBuffer` would only ever be called when the `$` is present in the name to begin with, so at least it isn't slowing down got-to-definition for all requests.

Reply:
> I guess the getBuffer would only ever be called when the $ is present in the name to begin with, so at least it isn't slowing down got-to-definition for all requests.

Yes, that's correct.

> Does exists() help ?

I think the answer is no. My understanding is that, as long as the object returns from `JavaCore.create` is not null, that means it's in the Java model, then `exists()` will always return true.

Considering the original request is not 'hot', maybe hang on this change a little bit? At least there is a solution.